contributors/design-proposals/admission_control_extension.md

```diff
@@ -0,0 +1,507 @@
+
+Once created, an object is not visible to clients unless the following conditions are satisfied:
+
+1. The initializers field is an empty array.
+2. The client provides a special option to GET, LIST, or WATCH indicating that the user wishes to see uninitialized objects.
+
+Each initializer is a controller or other client agent that watches for new objects with an initializer
+whose first position matches their assigned name (e.g. `PodAutoSizer`) and then operate on them. These
+clients would use the `?includeUninitializedObjects=true` query param (working name) and observe all
+objects.
+
+The initializer would perform a normal update on the object to perform their function, and then
+remove their entry from `initializers` (or adding more entries). If an error occurs during initialization
+that must terminate initialization, the `Status` field on the initializer should be set instead of removing
+the initializer entry and then the initializer should delete the object.
+
+During initialization, resources may have relaxed validation requirements, which means initializers must
+handle undefaulted or invalid objects. The update that completes initialization will be defaulted as
+normal and the normal validation logic should be enforced. UID and creation timestamp must be set before
+initializers run. The initial object creation must still be validated - resource types that wish to support
+deep initialization must design their objects to accept those.
+
+To allow naive clients to avoid having to deal with uninitialized objects, the API will automatically
+filter uninitialized objects out LIST and WATCH. Explicit GETs to that object should return the
+appropriate status code `202 Accepted` indicating that the resource is reserved and including a `Status`
+response with the correct resource version, but not the object. Clients specifying
+`includeUninitializedObjects` will see all updates, but shared code like caches and informers may need
+to implement layered filters to handle multiple clients requesting both variants. A CREATE to an
+uninitialized object should report the same status as before, and DELETE is always allowed.
+
+Finally, the apiserver that accepts the incoming creation should hold the response until the object is
+initialized or the timeout is exceeded. This increases latency, but allows clients to avoid breaking
+semantics. If the apiserver reaches the timeout it must return an appropriate error that includes the
+resource version of the object and UID so that clients can perform a watch. If an initializer reports
+a `Status` with a failure, it must return that to the user (all failures are correlated with deletion).
+
+There is no current error case for a timeout that exactly matches existing behavior except a 5xx
+timeout if etcd does not respond quickly. We should return that error, but return an appropriate
+cause that lets a client determine what the outcome was.
```

Comment:
We don't need it initially, but might want a TTL for resources that are never fully initialized.